### DISCUSSION TITLE
Store fonts and sprites in PMTiles file

### DISCUSSION THREAD

**Discussion by wipfli** [protomaps/PMTiles#141] [Ideas]
I think it would be super convenient if we could store fonts and sprites in `.pmtiles` files. Like that, all the stuff needed to make a MapLibre GL JS map would be in one place...

**Comment by bdon** [comment]
(Moved to discussion)

I get the motivation, but I think Styles themselves would be a better candidate for combining with a tileset, since fonts/sprites are only associated with a tileset through a style.

You can as-is store a JSON Style within the arbitrary JSON section of PMTiles.

You could even hack in a way to store fonts and sprites as base64 encoded values in there, but that's pretty ugly.

One thing I have thought about is doing a range-request like thing with `addProtocol` to fetch parts of a single font file instead of 256 font files, but that's only a reduction from 256 files to 1 file instead of the orders of magnitude of PMTiles (10,000,000 to 1 etc)

**Reply by wipfli** [reply]
You are right, styles are important too. Maybe pmtiles is the wrong place to store everything needed for a map, but I would just find it so convenient to have a tool which has as input some tiles, fonts, sprites, styles, and as output one file which I can put on whatever storage provider and then my frontend library of choice (MapLibre GL JS) goes to that one file and gets what it needs.

In particular this would be cool for demo. I already now enjoy that I can finally host self-contained demos with pmtiles. For example this one: https://wipfli.github.io/single-highway-layer/, but it would be even nicer if I could have the fonts bundled up too and and sprites...

**Reply by bdon** [reply]
Another thing to think about is that the fonts and sprites are usually the same across tilesets - it would be redundant for all tilesets to have duplicate copies of identical data. So the parting out of sprites, fonts, styles and tilesets into separate files still makes sense to me. I think the biggest opportunity is to rethink the font format to something more convenient (or get rid of PBF fonts completely)

**Reply by tyrauber** [reply]
@bdon, do you have an example of using sprites with PMTiles and MapLibre?  All I found was [this sprite example ](https://github.com/protomaps/protomaps.js/blob/main/examples/sprites.html) with leaflet. I'd like to do something similar with maplibre and a v2 pmtile. Any thoughts?

**Reply by bdon** [reply]
@tyrauber there isn't anything PMTiles specific related to Sprites - you'll need to create spritesheets and load them as typical for a MapLibre GL or Mapbox GL style. A standard style with some sprites is definitely on the roadmap for https://github.com/protomaps/basemaps - likely using the open icon sets available at the bottom of the page here: https://www.nextzen.org

**Comment by wipfli** [comment]
I am looping back to this idea because I think it would be so much easier for users of MapLibre if they just had to have a single file on their server, the `planet.pmtiles` file, and that would be the map basically. So that file would contain all the things needed to draw a map:

* vector tiles
* fonts
* sprites
* styles

The problem at the moment is that new users do not have a concept of styles, tiles, fonts, and sprites. They just want to get a map on their website. For that, they would probably expect to put some library (MapLibre GL JS and the pmtiles plugin) into their website and host the maps file on their server.

PMTiles simplified the tiles aspect of storing the map so much already. If we could also put the other stuff into the PMTiles file it would basically be perfect.

You are right @bdon that sprites and fonts can be reused between maps, but often a user just has a single map.

**Reply by bdon** [reply]
@wipfli seems like it could be a higher level project `map-starter-pack` that bundles together a pmtiles file, a folder of fonts, sprites, and styles, instead of extending PMTiles to do this. The fonts are going to be hundreds of files though :(

**Reply by wipfli** [reply]
Yes, a higher-level project seems the right approach to me.

**Comment by wipfli** [comment]
Esri has a vector tile package `.vtpk` format which essentially is what is discussed here I think...
